src/core_scan/Report_pro_findings.ml

```diff
@@ -0,0 +1,90 @@
+(*****************************************************************************)
+(* Purpose *)
+(*****************************************************************************)
+(** Support the reporting of pro findings
+
+  * Users want to know when a finding could only be found by the pro engine
+  * so that they can understand the value Semgrep provides. Additionally,
+  * we want to surface our "coolest" findings early to make sure users see
+  * them.
+
+  * To support this, we have a different type for the engine kind for matches,
+  * which includes a `PRO_SPECIFIC` option that reports when a finding could
+  * only be found using the pro engine. This option also allows us to highlight
+  * findings we think users should prioritize.
+  *)
+
+(*****************************************************************************)
+(* Helpers *)
+(*****************************************************************************)
+
+(* Check if the taint trace of the pattern crosses functions or files *)
+
+let tokens_of_trace_item (trace_item : Pattern_match.taint_trace_item) =
+  let rec tokens_of_call_trace (call_trace : Pattern_match.taint_call_trace) =
+    match call_trace with
+    | Toks ts -> ts
+    | Call { call_toks; intermediate_vars; call_trace } ->
```

Comment:
You don't need to look into all the `intermediate_vars`, one would be enough, all must belong to the same file.

Reply:
Didn't realize that, nice!